nanocoat/lib/scritchui/scritchPencilRasterShapes.c

```diff
@@ -17,7 +17,399 @@
+	sjme_scritchui_line* clipLine;
+	sjme_scritchui_pencilDrawPixelFunc drawPixel;
+
+
+	if (g == NULL)
+		return SJME_ERROR_NULL_ARGUMENTS;
+		
+	error = SJME_ERROR_NONE;
+
+	/* Java's coordinate system has positive angles moving counter-clockwise */
+	arcAngle = -arcAngle;
+	startAngle = -startAngle;
+
+	hasAlpha = g->hasAlpha;
+
+	drawPixel = g->prim.drawPixel;
+
+	/* Get clipping information. */
+	clipLine = &g->state.clipLine;
+
+	/** 
+	 * Only allocate the alpha buffer if the color isn't opaque. Noticeably
+	 * improves performance for opaque arcs.
+	 *
+	 * TODO: This can be optimized to take 1/8 of the usual memory size by
+	 * manipulating each bit of the boolean array below, as a _Bool variable
+	 * is often at least one whole byte in size.
+	 */
+	if(hasAlpha)
+	{
+		filledPixels = (sjme_jboolean *) sjme_alloca((w * h) *
+			sizeof(sjme_jboolean));
+		if (filledPixels == NULL)
+		{
+			error = sjme_error_outOfMemory(NULL, (w * h) *
+				sizeof(sjme_jboolean));
+			goto fail_any;
+		}
+		memset(filledPixels, 0, (w * h) * sizeof(sjme_jboolean));
+	}
+
+	/**
+	 * This is just drawArc's Bresenham midpoint circle algorithm modified to 
+	 * draw arcs from the center to the edge and check if a pixel was already
+	 * painted before. Works great but is slow on larger transparent arcs.
+	 * Trying to use a scanline approach could result in better performance due
+	 * to the simpler checks (just go from yMax to yMin while making sure a y 
+	 * position is never used twice) but doesn't seem to work right in some
+	 * edge cases (negative angles, too wide/narrow arc angles, etc) so this
+	 * one, albeit slow, is what we'll be going with at the moment.
+	 * 
+	 * At least it has the upside of generating far more stable arcs that fully
+	 * match their outline compared to Java AWT's algorithm.
+	 * 
+	 * TODO: Optimize this later
+	 */ 
+	
+	centerX = x + w / 2.0f;
+	centerY = y + h / 2.0f;
+	radiusX = w / 2.0f;
+	radiusY = h / 2.0f;
+	startAngleRad = startAngle * 0.017453292f;
+	endAngleRad = ((startAngle + arcAngle) * 0.017453292f) 
+	- startAngleRad;
+
+	maxRad = radiusX > radiusY ? radiusX : radiusY;
+
+	steps = fabs(arcAngle * ((w + h) / 2.0f) / 50.0f);
+	innerX, innerY;
+
+	for(int i = 0; i < steps; i++) 
+	{
+		angle = startAngleRad + (i * endAngleRad / steps);
+
+		for(int j = 0; j < maxRad; j++) 
+		{
+			innerX = round(centerX + radiusX * cos(angle) * (j / maxRad));
+			innerY = round(centerY + radiusY * sin(angle) * (j / maxRad));
+
+			/*
+			 * Make sure we're not drawing out of bounds. Or accessing the
+			 * alpha buffer at an invalid position with innerX-x or innerY-y
+			 */
+			if(innerX < clipLine->s.x || innerX >= clipLine->e.x
+				|| innerY < clipLine->s.y || innerY >= clipLine->e.y 
+				|| innerX-x < 0 || innerY-y < 0)
+				continue;
+
+			/** 
+			 * Only draw if opaque, or if the alpha buffer is not yet filled
+			 * for the current position.
+			 */
+			if(hasAlpha ? !*(filledPixels + ((innerY-y) * w + innerX-x)) : 1)
```

Comment:
`filledPixels[]` instead of `*filledPixels`, as it is more concise and readable.